File: src/robot_life/behavior/bt_nodes.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class NodeResult:
    node_name: str
    status: str
    details: str = ""


def run_node(node_name: str, behavior_id: str, degraded: bool = False) -> NodeResult:
    """Execute a minimal behavior-tree node with deterministic placeholder semantics."""
    if node_name == "state_idle":
        return NodeResult(node_name=node_name, status="success", details="idle")

    if node_name == "state_greet":
        return NodeResult(node_name=node_name, status="success", details="greet")

    if node_name == "state_attention":
        return NodeResult(node_name=node_name, status="success", details="attention")

    if node_name == "state_alert":
        return NodeResult(node_name=node_name, status="success", details="alert")

    if node_name == "state_observe":
        return NodeResult(node_name=node_name, status="success", details="observe")

    if node_name == "state_recover":
        return NodeResult(node_name=node_name, status="success", details="recover")

    if node_name == "guard_scene_validity":
        return NodeResult(node_name=node_name, status="success")

    if node_name == "act_nonverbal":
        detail = "degraded_nonverbal" if degraded else "full_nonverbal"
        return NodeResult(node_name=node_name, status="success", details=detail)

    if node_name == "act_speech_optional":
        detail = "speech_suppressed" if degraded else "speech_allowed"
        return NodeResult(node_name=node_name, status="success", details=detail)

    if node_name in {"monitor_preemption", "release"}:
        return NodeResult(node_name=node_name, status="success")

    return NodeResult(node_name=node_name, status="success", details=f"unknown_node:{behavior_id}")

File: Robot_Life_CPP/src/behavior/bt_nodes.cpp
#include "robot_life_cpp/behavior/bt_nodes.hpp"

#include <unordered_set>

namespace robot_life_cpp::behavior {

NodeResult run_node(const std::string& node_name, const std::string& behavior_id, bool degraded) {
  if (node_name == "state_idle") {
    return NodeResult{node_name, "success", "idle"};
  }
  if (node_name == "state_greet") {
    return NodeResult{node_name, "success", "greet"};
  }
  if (node_name == "state_attention") {
    return NodeResult{node_name, "success", "attention"};
  }
  if (node_name == "state_alert") {
    return NodeResult{node_name, "success", "alert"};
  }
  if (node_name == "state_observe") {
    return NodeResult{node_name, "success", "observe"};
  }
  if (node_name == "state_recover") {
    return NodeResult{node_name, "success", "recover"};
  }
  if (node_name == "guard_scene_validity") {
    return NodeResult{node_name, "success", ""};
  }
  if (node_name == "act_nonverbal") {
    return NodeResult{node_name, "success", degraded ? "degraded_nonverbal" : "full_nonverbal"};
  }
  if (node_name == "act_speech_optional") {
    return NodeResult{node_name, "success", degraded ? "speech_suppressed" : "speech_allowed"};
  }
  static const std::unordered_set<std::string> kSimpleSuccess{"monitor_preemption", "release"};
  if (kSimpleSuccess.contains(node_name)) {
    return NodeResult{node_name, "success", ""};
  }
  return NodeResult{node_name, "success", "unknown_node:" + behavior_id};
}

}  // namespace robot_life_cpp::behavior
